core/src/main/java/com/linecorp/armeria/internal/server/annotation/DecoratorUtil.java

```diff
@@ -0,0 +1,303 @@
+            if (udd != null) {
+                list.add(udd);
+                continue;
+            }
+
+            // If user-defined decorators are repeatable and they are specified more than once.
+            try {
+                final Method method = Iterables.getFirst(getMethods(annotation.annotationType(),
+                                                                    withName("value")), null);
+                assert method != null : "No 'value' method is found from " + annotation;
+                final Annotation[] decorators = (Annotation[]) method.invoke(annotation);
+                for (final Annotation decorator : decorators) {
+                    udd = userDefinedDecorator(decorator);
+                    if (udd == null) {
+                        break;
+                    }
+                    list.add(udd);
+                }
+            } catch (Throwable ignore) {
+                // The annotation may be a container of a decorator or may be not, so we just ignore
+                // any exception from this clause.
+            }
+        }
+    }
+
+    /**
+     * Returns a {@link HttpService} which is specified by {@link Decorator} annotations and user-defined
+     * decorator annotations.
+     */
+    public static HttpService applyDecorators(List<DecoratorAndOrder> decorators,
+                                              HttpService delegate) {
+        Function<? super HttpService, ? extends HttpService> decorator = Function.identity();
+        for (int i = decorators.size() - 1; i >= 0; i--) {
+            final DecoratorAndOrder d = decorators.get(i);
+            decorator = decorator.andThen(d.decorator());
+        }
+        return decorator.apply(delegate);
+    }
+
+    /**
+     * Returns a cached instance of the specified {@link Class} which is specified in the given
+     * {@link Annotation}.
+     */
+    static <T> T getInstance(Annotation annotation, Class<T> expectedType) {
+        try {
+            @SuppressWarnings("unchecked")
+            final Class<? extends T> clazz = (Class<? extends T>) invokeValueMethod(annotation);
+            return expectedType.cast(instanceCache.get(clazz));
+        } catch (ClassCastException e) {
+            throw new IllegalArgumentException(
+                    "A class specified in @" + annotation.annotationType().getSimpleName() +
+                    " annotation cannot be cast to " + expectedType, e);
+        }
+    }
+
+    private static <T> T getInstance0(Class<? extends T> clazz) throws Exception {
+        @SuppressWarnings("unchecked")
+        final Constructor<? extends T> constructor =
+                Iterables.getFirst(getConstructors(clazz, withParametersCount(0)), null);
+        assert constructor != null : "No default constructor is found from " + clazz.getName();
+        constructor.setAccessible(true);
+        return constructor.newInstance();
+    }
```

Comment:
How about extracting these methods into a utility class named `AnnotatedObjectFactory` so that 
`AnnotatedServiceFactory` and `DecoratorUtil` share code and the populated objects.